File: python/tracing/bedrock-openinference/bedrock_openinference_example.py
#!/usr/bin/env python3
"""
AWS Bedrock + OpenInference example for Respan tracing.

Uses openinference-instrumentation-bedrock to auto-instrument the boto3
Bedrock Runtime client and exports spans to Respan via respan-instrumentation-openinference.

Since we route LLM calls through the Respan gateway (OpenAI format),
a local HTTP proxy translates Bedrock Converse API requests to OpenAI format and back.
Dummy AWS credentials are used since calls never reach AWS.
"""
import json
import os
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path

import requests as http_requests
from dotenv import load_dotenv

ROOT_ENV_PATH = Path(__file__).resolve().parents[3] / ".env"
load_dotenv(ROOT_ENV_PATH, override=True)

RESPAN_API_KEY = os.getenv("RESPAN_GATEWAY_API_KEY") or os.getenv("RESPAN_API_KEY")
RESPAN_BASE_URL = (
    os.getenv("RESPAN_GATEWAY_BASE_URL")
    or os.getenv("RESPAN_BASE_URL")
    or "https://api.respan.ai/api"
).rstrip("/")

from openinference.instrumentation.bedrock import BedrockInstrumentor
from respan_instrumentation_openinference import OpenInferenceInstrumentor
from respan_tracing import RespanTelemetry
from respan_tracing.decorators import workflow

telemetry = RespanTelemetry(
    app_name="bedrock-openinference-example",
    api_key=RESPAN_API_KEY,
    is_batching_enabled=False,
    instruments=set(),
)

bedrock_oi = OpenInferenceInstrumentor(BedrockInstrumentor)
bedrock_oi.activate()


class BedrockConverseProxy(BaseHTTPRequestHandler):
    """Translates Bedrock Converse API requests to OpenAI chat completions.

    Botocore sends requests to paths like /model/{modelId}/converse.
    """

    def log_message(self, format, *args):
        pass

    def do_POST(self):
        content_length = int(self.headers.get("Content-Length", 0))
        body = json.loads(self.rfile.read(content_length)) if content_length else {}

        messages = []
        system_parts = body.get("system", [])
        if system_parts:
            sys_text = " ".join(p.get("text", "") for p in system_parts if "text" in p)
            if sys_text:
                messages.append({"role": "system", "content": sys_text})

        for msg in body.get("messages", []):
            role = msg.get("role", "user")
            content_parts = msg.get("content", [])
            text = " ".join(p.get("text", "") for p in content_parts if "text" in p)
            if text:
                messages.append({"role": role, "content": text})

        inference_config = body.get("inferenceConfig", {})
        openai_body = {
            "model": "gpt-4o-mini",
            "messages": messages,
            "temperature": inference_config.get("temperature", 0),
            "max_tokens": inference_config.get("maxTokens", 256),
        }

        resp = http_requests.post(
            f"{RESPAN_BASE_URL}/chat/completions",
            json=openai_body,
            headers={
                "Authorization": f"Bearer {RESPAN_API_KEY}",
                "Content-Type": "application/json",
            },
        )

        if resp.status_code != 200:
            self._send_json(resp.status_code, json.loads(resp.content))
            return

        openai_resp = resp.json()
        choice = openai_resp.get("choices", [{}])[0]
        usage = openai_resp.get("usage", {})

        bedrock_resp = {
            "output": {
                "message": {
                    "role": "assistant",
                    "content": [{"text": choice.get("message", {}).get("content", "")}],
                }
            },
            "stopReason": "end_turn",
            "usage": {
                "inputTokens": usage.get("prompt_tokens", 0),
                "outputTokens": usage.get("completion_tokens", 0),
                "totalTokens": usage.get("total_tokens", 0),
            },
            "metrics": {"latencyMs": 100},
        }

        self._send_json(200, bedrock_resp)

    def _send_json(self, status, data):
        payload = json.dumps(data).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)


def start_proxy():
    server = HTTPServer(("127.0.0.1", 0), BedrockConverseProxy)
    port = server.server_address[1]
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    return server, f"http://127.0.0.1:{port}"


proxy_server, proxy_url = start_proxy()

import boto3

client = boto3.client(
    "bedrock-runtime",
    region_name="us-east-1",
    endpoint_url=proxy_url,
    aws_access_key_id="dummy",
    aws_secret_access_key="dummy",
)


@workflow(name="bedrock_openinference_workflow")
def bedrock_workflow(topic: str) -> str:
    response = client.converse(
        modelId="anthropic.claude-3-haiku-20240307-v1:0",
        messages=[
            {
                "role": "user",
                "content": [
                    {"text": f"Write a haiku about {topic}. Output ONLY the haiku."}
                ],
            }
        ],
        inferenceConfig={"temperature": 0, "maxTokens": 256},
    )

    output_message = response["output"]["message"]
    return output_message["content"][0]["text"]


def main() -> None:
    print("=" * 60)
    print("Bedrock OpenInference Example")
    print("=" * 60)

    if not RESPAN_API_KEY:
        print("Skipping: RESPAN_API_KEY not set.")
        return

    try:
        result = bedrock_workflow(topic="recursion in programming")
        print(result)
    finally:
        telemetry.flush()
        bedrock_oi.deactivate()
        proxy_server.shutdown()


if __name__ == "__main__":
    main()

File: typescript/tracing/bedrock-openinference/bedrock_openinference_example.ts
/**
 * AWS Bedrock — Chat completion with Respan tracing via OpenInference.
 *
 * The @aws-sdk/client-bedrock-runtime SDK sends requests to
 *   https://bedrock-runtime.{region}.amazonaws.com/model/{modelId}/converse
 *
 * Since we route LLM calls through the Respan gateway (OpenAI format),
 * a local HTTP proxy translates the Bedrock Converse API to OpenAI and back.
 *
 * The @arizeai/openinference-instrumentation-bedrock patches
 * BedrockRuntimeClient.prototype.send to create OpenInference-format spans.
 */

import "dotenv/config";
import http from "node:http";
import https from "node:https";
import { URL } from "node:url";
import { trace } from "@opentelemetry/api";
import { BedrockInstrumentation } from "@arizeai/openinference-instrumentation-bedrock";
import * as BedrockModule from "@aws-sdk/client-bedrock-runtime";
import {
  BedrockRuntimeClient,
  ConverseCommand,
} from "@aws-sdk/client-bedrock-runtime";
import { Respan } from "@respan/respan";
import { OpenInferenceInstrumentor } from "@respan/instrumentation-openinference";

const RESPAN_API_KEY = process.env.RESPAN_API_KEY!;
const RESPAN_BASE_URL = (
  process.env.RESPAN_BASE_URL ?? "https://api.respan.ai/api"
).replace(/\/+$/, "");

// ── 1. Initialize Respan tracing with OpenInference instrumentor ──────
const respan = new Respan({
  apiKey: RESPAN_API_KEY,
  baseURL: RESPAN_BASE_URL,
  instrumentations: [
    new OpenInferenceInstrumentor(BedrockInstrumentation),
  ],
});
await respan.initialize();

// ── 2. Manually instrument @aws-sdk/client-bedrock-runtime ────────────
const bedrockInst = new BedrockInstrumentation();
bedrockInst.setTracerProvider(trace.getTracerProvider());
bedrockInst.manuallyInstrument(BedrockModule);

// ── 3. Local proxy: Bedrock Converse → OpenAI → Respan gateway ───────
function startProxy(targetBase: string, apiKey: string): Promise<string> {
  return new Promise((resolve) => {
    const server = http.createServer((req, res) => {
      const chunks: Buffer[] = [];
      req.on("data", (c: Buffer) => chunks.push(c));
      req.on("end", () => {
        try {
          const body = JSON.parse(Buffer.concat(chunks).toString());

          // Convert Bedrock Converse messages → OpenAI messages
          const messages = (body.messages ?? []).map(
            (m: { role: string; content: { text: string }[] }) => ({
              role: m.role === "assistant" ? "assistant" : "user",
              content:
                m.content
                  ?.map((c: { text?: string }) => c.text)
                  .filter(Boolean)
                  .join("") ?? "",
            })
          );

          // Add system message if present
          if (body.system?.length) {
            messages.unshift({
              role: "system",
              content: body.system
                .map((s: { text?: string }) => s.text)
                .filter(Boolean)
                .join("\n"),
            });
          }

          const openaiBody = JSON.stringify({
            model: "gpt-4o-mini",
            messages,
            temperature:
              body.inferenceConfig?.temperature ?? 0,
            max_tokens:
              body.inferenceConfig?.maxTokens ?? 256,
          });

          const target = new URL(`${targetBase}/chat/completions`);
          const options: https.RequestOptions = {
            hostname: target.hostname,
            port:
              target.port ||
              (target.protocol === "https:" ? 443 : 80),
            path: target.pathname,
            method: "POST",
            headers: {
              "Content-Type": "application/json",
              "Content-Length": Buffer.byteLength(openaiBody),
              Authorization: `Bearer ${apiKey}`,
            },
          };

          const transport =
            target.protocol === "https:" ? https : http;
          const upstream = transport.request(options, (upRes) => {
            const respChunks: Buffer[] = [];
            upRes.on("data", (c: Buffer) => respChunks.push(c));
            upRes.on("end", () => {
              try {
                const openaiResp = JSON.parse(
                  Buffer.concat(respChunks).toString()
                );
                const choice = openaiResp.choices?.[0];

                // Convert OpenAI response → Bedrock Converse format
                const bedrockResp = {
                  output: {
                    message: {
                      role: "assistant",
                      content: [
                        {
                          text: choice?.message?.content ?? "",
                        },
                      ],
                    },
                  },
                  stopReason: "end_turn",
                  usage: {
                    inputTokens:
                      openaiResp.usage?.prompt_tokens ?? 0,
                    outputTokens:
                      openaiResp.usage?.completion_tokens ?? 0,
                    totalTokens:
                      openaiResp.usage?.total_tokens ?? 0,
                  },
                  metrics: {
                    latencyMs: 0,
                  },
                };

                const payload = JSON.stringify(bedrockResp);
                res.writeHead(200, {
                  "Content-Type": "application/json",
                });
                res.end(payload);
              } catch {
                res.writeHead(502);
                res.end(
                  JSON.stringify({
                    error: "Failed to parse upstream response",
                  })
                );
              }
            });
          });
          upstream.on("error", (e) => {
            res.writeHead(502);
            res.end(JSON.stringify({ error: e.message }));
          });
          upstream.write(openaiBody);
          upstream.end();
        } catch {
          res.writeHead(400);
          res.end(JSON.stringify({ error: "Bad request" }));
        }
      });
    });

    server.listen(0, () => {
      const port = (server.address() as any).port;
      (globalThis as any).__bedrockProxy = server;
      resolve(`http://localhost:${port}`);
    });
  });
}

const proxyUrl = await startProxy(RESPAN_BASE_URL, RESPAN_API_KEY);

// ── 4. Create Bedrock client pointing at local proxy ──────────────────
const client = new BedrockRuntimeClient({
  region: "us-east-1",
  endpoint: proxyUrl,
  credentials: {
    accessKeyId: "proxy-passthrough",
    secretAccessKey: "proxy-passthrough",
  },
});

const command = new ConverseCommand({
  modelId: "anthropic.claude-3-haiku-20240307-v1:0",
  messages: [
    {
      role: "user",
      content: [
        { text: "Write a haiku about recursion in programming." },
      ],
    },
  ],
  inferenceConfig: {
    temperature: 0,
    maxTokens: 256,
  },
});

const result = await client.send(command);

const responseText = result.output?.message?.content
  ?.map((c) => c.text)
  .join("");
console.log(responseText);

// ── 5. Flush and cleanup ──────────────────────────────────────────────
await respan.flush();
((globalThis as any).__bedrockProxy as http.Server).close();
